### PR TITLE
Add negative stateapp cases to e2e tests

### PR DESCRIPTION
**Pull request by halspang**
# Description

Negative test cases include:

 - HTTP content limits
 - Missing key
 - Missing statestore

## Issue reference

closes https://github.com/dapr/dapr/issues/1810

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing (stateapp)
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: tests/e2e/stateapp/stateapp_test.go [hunk 1]
```diff
@@ -263,6 +269,50 @@ func generateTestCases(isHTTP bool) []testCase {
...
+		},
+		{
+			// No comma since this will become the name of the test without spaces.
+			"Test missing key handled " + protocol,
+			[]testStep{
+				{
+					"get",
+					newRequest(utils.SimpleKeyValue{testCase1Key, nil}),
+					emptyResponse,
+					200,
```
Comment (mukundansundar) [inline tests/e2e/stateapp/stateapp_test.go:311]: shouldn't missing key be 404?

File: tests/e2e/stateapp/stateapp_test.go [hunk 2]
```diff
@@ -415,3 +481,47 @@ func TestStateTransactionApps(t *testing.T) {
...
+	_, err := utils.HTTPGetNTimes(externalURL, numHealthChecks)
+	require.NoError(t, err)
+
+	// Define state URL for statestore that does not exist.
+	stateURL := "http://localhost:3500/v1.0/state/nonexistantstate"
+
+	request := newRequest(utils.SimpleKeyValue{guuid.New().String(), nil})
+	body, _ := json.Marshal(request)
+
+	resp, err := utils.HTTPPost(stateURL, body)
```
Comment (mukundansundar) [inline tests/e2e/stateapp/stateapp_test.go:500]: use PostWithStatus to get the status and verify the status also.

File: tests/e2e/stateapp/stateapp_test.go [hunk 3]
```diff
@@ -415,3 +481,47 @@ func TestStateTransactionApps(t *testing.T) {
...
+	require.NoError(t, err)
+
+	// Define state URL for statestore that does not exist.
+	for _, store := range []string{"badhost-store", "badpass-store"} {
+		stateURL := fmt.Sprintf("http://localhost:3500/v1.0/state/%v", store)
+
+		request := newRequest(utils.SimpleKeyValue{guuid.New().String(), nil})
+		body, _ := json.Marshal(request)
+
+		resp, err := utils.HTTPPost(stateURL, body)
```
Comment (mukundansundar) [inline tests/e2e/stateapp/stateapp_test.go:522]: use PostWithStatus to verify status code.

**Review by mukundansundar — CHANGES_REQUESTED** [review]

File: tests/e2e/stateapp/stateapp_test.go (hunk 1 above)
Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go:311]: The behavior I've observed has been a missing key gave empty state though that could be the test app eating the error. I'll send a request directly to the statestore and verify that.

File: tests/e2e/stateapp/stateapp_test.go (hunk 2 above)
Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go]: Done, though it appears that the state app eats a lot of error messages. You can see it in the response strings though.

File: tests/e2e/stateapp/stateapp_test.go (hunk 3 above)
Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go]: Done, though it appears that the state app eats a lot of error messages. You can see it in the response strings though.

File: tests/e2e/stateapp/stateapp_test.go (hunk 1 above)
Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go]: Looking through how everything works, this does appear to be the expected behavior. I've separated it into a separate test that makes it a bit clearer but still no 404.

File: tests/e2e/stateapp/stateapp_test.go [hunk 4]
```diff
@@ -263,6 +263,37 @@ func generateTestCases(isHTTP bool) []testCase {
...
+					"save",
+					generateSpecificLengthSample(1024*1024*3 - 55), // Exact limit
+					emptyResponse,
+					201,
+				},
+				{
+					"save",
+					generateSpecificLengthSample(1024*1024*4 + 1), // Limit + 1
+					emptyResponse,
+					500,
```
Comment (mukundansundar) [inline tests/e2e/stateapp/stateapp_test.go:286]: shouldn't this be 201 for gRPC?

File: tests/e2e/stateapp/stateapp_test.go [hunk 5]
```diff
@@ -415,3 +462,98 @@ func TestStateTransactionApps(t *testing.T) {
...
+func TestMissingKeyDirect(t *testing.T) {
+	externalURL := tr.Platform.AcquireAppExternalURL(appName)
+	require.NotEmpty(t, externalURL, "external URL must not be empty!")
+
+	// This initial probe makes the test wait a little bit longer when needed,
+	// making this test less flaky due to delays in the deployment.
+	_, err := utils.HTTPGetNTimes(externalURL, numHealthChecks)
+	require.NoError(t, err)
+
+	for _, protocol := range []string{"http", "grpc"} {
```
Comment (mukundansundar) [inline tests/e2e/stateapp/stateapp_test.go:475]: use `t.Run(name, func())` here?

**Review by mukundansundar — COMMENTED** [review]
use t.Run subtest for missing misconfigured and missingkey tests.

File: tests/e2e/stateapp/stateapp_test.go (hunk 4 above)
Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go:286]: That's what I thought at first but the state app was actually reseting the error code. So it appeared that gRPC didn't have the same limitations but it in fact does.

File: tests/e2e/stateapp/stateapp_test.go (hunk 5 above)
Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go:475]: Done.

File: tests/e2e/stateapp/stateapp_test.go [hunk 6]
```diff
@@ -415,3 +462,103 @@ func TestStateTransactionApps(t *testing.T) {
...
+			key := guuid.New().String()
+			request := newRequest(utils.SimpleKeyValue{key, nil})
+			body, _ := json.Marshal(request)
+
+			resp, status, err := utils.HTTPPostWithStatus(stateURL, body)
+
+			var states requestResponse
+			json.Unmarshal(resp, &states)
+
+			require.Equal(t, 200, status)
```
Comment (mukundansundar) [inline tests/e2e/stateapp/stateapp_test.go:488]: shouldn't the response be no content(204) instead of ok(200)?

Comment (halspang) [reply tests/e2e/stateapp/stateapp_test.go:488]: 204 is a no content and this technically did return some content. It's a JSON blob with a key that references no value. So it doesn't really mean much, but by status code standards I think is correct.

**Review by mukundansundar — DISMISSED** [review]
lgtm

@wcs1only @yaron2  Can you please comment on https://github.com/dapr/dapr/pull/2329#discussion_r514605560?

File: tests/e2e/stateapp/stateapp_test.go (hunk 6 above)
Comment (wcs1only) [reply tests/e2e/stateapp/stateapp_test.go:488]: I have this question asked in #2338. It's fine to ship this as is. If I end up changing the code to give a 204 here, I'll circle back and fix this test.

**Review by vinayada1 — APPROVED** [review]

**Pull request merged**